# Files

## Persistence

So far, we have learned how to write programs and communicate our intentions to the *Central Processing Unit* using conditional execution, functions, and iterations. We have learned how to create and use data structures in the *Main Memory*. The CPU and memory are where our software works and runs. It is where all of the "thinking" happens.

But once the power is turned off, anything stored in either the CPU or main memory is erased. So up to now, our programs have just been transient fun exercises to learn Python.

In this chapter, we start to work with *Secondary Memory* (or files). Secondary memory is not erased when the power is turned off. Or in the case of a USB flash drive, the data we write from our programs can be removed from the system and transported to another system.

We will primarily focus on reading and writing text files such as those we create in a text editor. Later we will see how to work with database files which are binary files, specifically designed to be read and written through database software.

## Opening files

When we want to read or write a file (say on your hard drive), we first must *open* the file. Opening the file communicates with your operating system, which knows where the data for each file is stored. When you open a file, you are asking the operating system to find the file by name and make sure the file exists. 

In this example, we open the file *mbox.txt*, which should be stored in the same folder that you are in when you start Python. You can download this file from www.edu.susu.ru and save it in your Jupyter folder. Save also *mbox-short.txt*.

In [2]:
fhand = open('mbox.txt')
print(fhand)

<_io.TextIOWrapper name='mbox.txt' mode='r' encoding='cp1251'>


If the __open__ is successful, the operating system returns us a *file handle*. The file handle is not the actual data contained in the file, but instead it is a "handle" that we can use to read the data. You are given a handle if the requested file exists and you have the proper permissions to read the file.

If the file does not exist, __open__ will fail with a traceback and you will not get a handle to access the contents of the file:

In [3]:
fhand = open('stuff.txt')

FileNotFoundError: [Errno 2] No such file or directory: 'stuff.txt'

Later we will use __try__ and __except__ to deal more gracefully with the situation where we attempt to open a file that does not exist.

## Text files and lines

A text file can be thought of as a sequence of lines, much like a Python string can be thought of as a sequence of characters. For example, this is a sample of a text file *mbox.txt* which records mail activity from various individuals in an open source project development team:

From stephen.marquard@uct.ac.za Sat Jan  5 09:14:16 2008
Return-Path: <postmaster@collab.sakaiproject.org>
Date: Sat, 5 Jan 2008 09:12:18 -0500
To: source@collab.sakaiproject.org
From: stephen.marquard@uct.ac.za
Subject: [sakai] svn commit: r39772 - content/branches/
Details: http://source.sakaiproject.org/viewsvn/?view=rev&rev=39772
...

The file is in a standard format for a file containing multiple mail messages. The lines which start with "From " separate the messages and the lines which start with "From:" are part of the messages. For more information about the mbox format, see https://en.wikipedia.org/wiki/Mbox.

To break the file into lines, there is a special character that represents the "end of the line" called the *newline* character.

In Python, we represent the *newline* character as a backslash-n in string constants. Even though this looks like two characters, it is actually a single character. When we look at the variable by entering "stuff" in the interpreter, it shows us the __\n__ in the string, but when we use print to show the string, we see the string broken into two lines by the newline character.

In [5]:
stuff = 'Hello\nWorld!'
stuff

'Hello\nWorld!'

In [6]:
print(stuff)

Hello
World!


In [7]:
stuff = 'X\nY'
print(stuff)

X
Y


In [8]:
len(stuff)

3

You can also see that the length of the string __X\nY__ is three characters because the newline character is a single character.

So when we look at the lines in a file, we need to *imagine* that there is a special invisible character called the newline at the end of each line that marks the end of the line.

So the newline character separates the characters in the file into lines.

## Reading files

While the *file handle* does not contain the data for the file, it is quite easy to construct a __for__ loop to read through and count each of the lines in a file:

In [9]:
fhand = open('mbox.txt')
count = 0
for line in fhand:
    count = count + 1
print('Line Count:', count)

Line Count: 132045


We can use the file handle as the sequence in our __for__ loop. Our __for__ loop simply counts the number of lines in the file and prints them out. The rough translation of the __for__ loop into English is, "for each line in the file represented by the file handle, add one to the __count__ variable."

The reason that the __open__ function does not read the entire file is that the file might be quite large with many gigabytes of data. The __open__ statement takes the same amount of time regardless of the size of the file. The __for__ loop actually causes the data to be read from the file.

When the file is read using a __for__ loop in this manner, Python takes care of splitting the data in the file into separate lines using the newline character. Python reads each line through the newline and includes the newline as the last character in the __line__ variable for each iteration of the __for__ loop.

Because the __for__ loop reads the data one line at a time, it can efficiently read and count the lines in very large files without running out of main memory to store the data. The above program can count the lines in any size file using very little memory since each line is read, counted, and then discarded.

If you know the file is relatively small compared to the size of your main memory, you can read the whole file into one string using the __read__ method on the file handle.

In [11]:
fhand = open('mbox.txt')
inp = fhand.read()
print(len(inp))

6687002


In [12]:
print(inp[:20])

From stephen.marquar


In this example, the entire contents (all 6,687,002 characters) of the file *mbox.txt* are read directly into the variable __inp__. We use string slicing to print out the first 20 characters of the string data stored in __inp__.

When the file is read in this manner, all the characters including all of the lines and newline characters are one big string in the variable __inp__. It is a good idea to store the output of __read__ as a variable because each call to __read__ exhausts the resource:

In [17]:
fhand = open('mbox.txt')
print(len(fhand.read()))

6687002


In [18]:
print(len(fhand.read()))

0


Remember that this form of the __open__ function should only be used if the file data will fit comfortably in the main memory of your computer. If the file is too large to fit in main memory, you should write your program to read the file in chunks using a __for__ or __while__ loop.

## Searching through a file

When you are searching through data in a file, it is a very common pattern to read through a file, ignoring most of the lines and only processing lines which meet a particular condition. We can combine the pattern for reading a file with string methods to build simple search mechanisms.

For example, if we wanted to read a file and only print out lines which started with the prefix "From:", we could use the string method *startswith* to select only those lines with the desired prefix:

In [23]:
fhand = open('mbox-short.txt')
for line in fhand:
    if line.startswith('From:'):
        print(line)

From: stephen.marquard@uct.ac.za

From: louis@media.berkeley.edu

From: zqian@umich.edu

From: rjlowe@iupui.edu

From: zqian@umich.edu

From: rjlowe@iupui.edu

From: cwen@iupui.edu

From: cwen@iupui.edu

From: gsilver@umich.edu

From: gsilver@umich.edu

From: zqian@umich.edu

From: gsilver@umich.edu

From: wagnermr@iupui.edu

From: zqian@umich.edu

From: antranig@caret.cam.ac.uk

From: gopal.ramasammycook@gmail.com

From: david.horwitz@uct.ac.za

From: david.horwitz@uct.ac.za

From: david.horwitz@uct.ac.za

From: david.horwitz@uct.ac.za

From: stephen.marquard@uct.ac.za

From: louis@media.berkeley.edu

From: louis@media.berkeley.edu

From: ray@media.berkeley.edu

From: cwen@iupui.edu

From: cwen@iupui.edu

From: cwen@iupui.edu



The output looks great since the only lines we are seeing are those which start with "From:", but why are we seeing the extra blank lines? This is due to that invisible *newline* character. Each of the lines ends with a newline, so the __print__ statement prints the string in the variable *line* which includes a newline and then __print__ adds *another* newline, resulting in the double spacing effect we see.

We could use line slicing to print all but the last character, but a simpler approach is to use the *rstrip* method which strips whitespace from the right side of a string as follows:

In [22]:
fhand = open('mbox-short.txt')
for line in fhand:
    line = line.rstrip()
    if line.startswith('From:'):
        print(line)

From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: gsilver@umich.edu
From: gsilver@umich.edu
From: zqian@umich.edu
From: gsilver@umich.edu
From: wagnermr@iupui.edu
From: zqian@umich.edu
From: antranig@caret.cam.ac.uk
From: gopal.ramasammycook@gmail.com
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: louis@media.berkeley.edu
From: ray@media.berkeley.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu


As your file processing programs get more complicated, you may want to structure your search loops using __continue__. The basic idea of the search loop is that you are looking for "interesting" lines and effectively skipping "uninteresting" lines. And then when we find an interesting line, we do something with that line.

We can structure the loop to follow the pattern of skipping uninteresting lines as follows:

In [24]:
fhand = open('mbox-short.txt')
for line in fhand:
    line = line.rstrip()
    # Skip 'uninteresting lines'
    if not line.startswith('From:'):
        continue
    # Process our 'interesting' line
    print(line)

From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: gsilver@umich.edu
From: gsilver@umich.edu
From: zqian@umich.edu
From: gsilver@umich.edu
From: wagnermr@iupui.edu
From: zqian@umich.edu
From: antranig@caret.cam.ac.uk
From: gopal.ramasammycook@gmail.com
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: louis@media.berkeley.edu
From: ray@media.berkeley.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu


The output of the program is the same. In English, the uninteresting lines are those which do not start with "From:", which we skip using __continue__. For the "interesting" lines (i.e., those that start with "From:") we perform the processing on those lines.

We can use the __find__ string method to simulate a text editor search that finds lines where the search string is anywhere in the line. Since __find__ looks for an occurrence of a string within another string and either returns the position of the string or -1 if the string was not found, we can write the following loop to show lines which contain the string "@uct.ac.za" (i.e., they come from the University of Cape Town in South Africa):

In [25]:
fhand = open('mbox-short.txt')
for line in fhand:
    line = line.rstrip()
    if line.find('@uct.ac.za') == -1: continue
    print(line)

From stephen.marquard@uct.ac.za Sat Jan  5 09:14:16 2008
X-Authentication-Warning: nakamura.uits.iupui.edu: apache set sender to stephen.marquard@uct.ac.za using -f
From: stephen.marquard@uct.ac.za
Author: stephen.marquard@uct.ac.za
From david.horwitz@uct.ac.za Fri Jan  4 07:02:32 2008
X-Authentication-Warning: nakamura.uits.iupui.edu: apache set sender to david.horwitz@uct.ac.za using -f
From: david.horwitz@uct.ac.za
Author: david.horwitz@uct.ac.za
r39753 | david.horwitz@uct.ac.za | 2008-01-04 13:05:51 +0200 (Fri, 04 Jan 2008) | 1 line
From david.horwitz@uct.ac.za Fri Jan  4 06:08:27 2008
X-Authentication-Warning: nakamura.uits.iupui.edu: apache set sender to david.horwitz@uct.ac.za using -f
From: david.horwitz@uct.ac.za
Author: david.horwitz@uct.ac.za
From david.horwitz@uct.ac.za Fri Jan  4 04:49:08 2008
X-Authentication-Warning: nakamura.uits.iupui.edu: apache set sender to david.horwitz@uct.ac.za using -f
From: david.horwitz@uct.ac.za
Author: david.horwitz@uct.ac.za
From david.horw

Here we also use the contracted form of the __if__ statement where we put the __continue__ on the same line as the __if__. This contracted form of the __if__ functions the same as if the __continue__ were on the next line and indented.

## Letting the user choose the file name

We really do not want to have to edit our Python code every time we want to process a different file. It would be more usable to ask the user to enter the file name string each time the program runs so they can use our program on different files without changing the Python code.

This is quite simple to do by reading the file name from the user using __input__ as follows:

In [26]:
fname = input('Enter the file name: ')
fhand = open(fname)
count = 0
for line in fhand:
    if line.startswith('Subject:'):
        count = count + 1
print('There were', count, 'subject lines in', fname)

Enter the file name: mbox.txt
There were 1797 subject lines in mbox.txt


We read the file name from the user and place it in a variable named __fname__ and open that file. Now we can run the program repeatedly on different files.

Before peeking at the next section, take a look at the above program and ask yourself, "What could go possibly wrong here?" or "What might our friendly user do that would cause our nice little program to ungracefully exit with a traceback, making us look not-so-cool in the eyes of our users?"

## Using try, except, and open

I told you not to peek. This is your last chance.

What if our user types something that is not a file name?

In [27]:
fname = input('Enter the file name: ')
fhand = open(fname)
count = 0
for line in fhand:
    if line.startswith('Subject:'):
        count = count + 1
print('There were', count, 'subject lines in', fname)

Enter the file name: missing.txt


FileNotFoundError: [Errno 2] No such file or directory: 'missing.txt'

Do not laugh. Users will eventually do every possible thing they can do to break your programs, either on purpose or with malicious intent. As a matter of fact, an important part of any software development team is a person or group called *Quality Assurance* (or QA for short) whose very job it is to do the craziest things possible in an attempt to break the software that the programmer has created.

The QA team is responsible for finding the flaws in programs before we have delivered the program to the end users who may be purchasing the software or paying our salary to write the software. So the QA team is the programmer's best friend.

So now that we see the flaw in the program, we can elegantly fix it using the __try/except__ structure. We need to assume that the __open__ call might fail and add recovery code when the __open__ fails as follows:

In [29]:
fname = input('Enter the file name: ')
try:
    fhand = open(fname)
except:
    print('File cannot be opened:', fname)
    exit()
count = 0
for line in fhand:
    if line.startswith('Subject:'):
        count = count + 1
print('There were', count, 'subject lines in', fname)

Enter the file name: missing.txt
File cannot be opened: missing.txt
There were 0 subject lines in missing.txt


The __exit__ function terminates the program. It is a function that we call that never returns. Now when our user (or QA team) types in silliness or bad file names, we "catch" them and recover gracefully:

Protecting the __open__ call is a good example of the proper use of __try__ and __except__ in a Python program. We use the term "Pythonic" when we are doing something the "Python way". We might say that the above example is the Pythonic way to open a file.

Once you become more skilled in Python, you can engage in repartee with other Python programmers to decide which of two equivalent solutions to a problem is "more Pythonic". The goal to be "more Pythonic" captures the notion that programming is part engineering and part art. We are not always interested in just making something work, we also want our solution to be elegant and to be appreciated as elegant by our peers.

## Writing files

To write a file, you have to open it with mode "w" as a second parameter:

In [1]:
fout = open('output.txt', 'w')
print(fout)

<_io.TextIOWrapper name='output.txt' mode='w' encoding='cp1251'>


If the file already exists, opening it in write mode clears out the old data and starts fresh, so be careful! If the file doesn't exist, a new one is created.

The __write__ method of the file handle object puts data into the file, returning the number of characters written. The default write mode is text for writing (and reading) strings.

In [2]:
line1 = "This here's the wattle,\n"
fout.write(line1)

24

Again, the file object keeps track of where it is, so if you call __write__ again, it adds the new data to the end.

We must make sure to manage the ends of lines as we write to the file by explicitly inserting the newline character when we want to end a line. The __print__ statement automatically appends a newline, but the __write__ method does not add the newline automatically.

In [3]:
line2 = 'the emblem of our land.\n'
fout.write(line2)

24

When you are done writing, you have to close the file to make sure that the last bit of data is physically written to the disk so it will not be lost if the power goes off.

In [4]:
fout.close()

We could close the files which we open for read as well, but we can be a little sloppy if we are only opening a few files since Python makes sure that all open files are closed when the program ends. When we are writing files, we want to explicitly close the files so as to leave nothing to chance.

## Debugging

When you are reading and writing files, you might run into problems with whitespace. These errors can be hard to debug because spaces, tabs, and newlines are normally invisible:

In [5]:
s = '1 2\t 3\n 4'
print(s)

1 2	 3
 4


The built-in function __repr__ can help. It takes any object as an argument and returns a string representation of the object. For strings, it represents whitespace characters with backslash sequences:

In [6]:
print(repr(s))

'1 2\t 3\n 4'


This can be helpful for debugging.

One other problem you might run into is that different systems use different characters to indicate the end of a line. Some systems use a newline, represented __\n__. Others use a return character, represented __\r__. Some use both. If you move files between different systems, these inconsistencies might cause problems.

For most systems, there are applications to convert from one format to another. You can find them (and read more about this issue) at wikipedia.org/wiki/Newline. Or, of course, you could write one yourself.

## Glossary

__catch__

To prevent an exception from terminating a program using the try and except statements. 

__newline__

A special character used in files and strings to indicate the end of a line. 

__Pythonic__

A technique that works elegantly in Python. "Using try and except is the Pythonic way to recover from missing files". 

__Quality Assurance__

A person or team focused on insuring the overall quality of a software product. QA is often involved in testing a product and identifying problems before the product is released. 

__text file__

A sequence of characters stored in permanent storage like a hard drive. 